packaging/manager.go

```diff
@@ -0,0 +1,148 @@
+)
+
+var logger = loggo.GetLogger("juju.packaging")
+
+// PackageManagerName describes a package manager.
+type PackageManagerName string
+
+// The list of supported package managers.
+const (
+	AptPackageManager    PackageManagerName = "apt"
+	YumPackageManager                       = "yum"
+	ZypperPackageManager                    = "zypper"
+	SnapPackageManager                      = "snap"
+)
+
+// Dependency is implemented by objects that can provide a series-specific
+// list of packages for installing a particular software dependency.
+type Dependency interface {
+	PackageList(series string) ([]Package, error)
+}
+
+// Package encapsulates the information required for installing a package.
+type Package struct {
+	// The name of the package to install
+	Name string
+
+	// Additional options to be passed to the package manager.
+	InstallOptions string
+
+	// The package manager to use for installing
+	PackageManager PackageManagerName
+}
+
+func (p *Package) appendInstallOptions(opt ...string) {
+	p.InstallOptions = strings.TrimSpace(
+		fmt.Sprintf("%s %s", p.InstallOptions, strings.Join(opt, " ")),
+	)
+}
+
+// MakePackageList returns a list of Package instances for each provided
+// package name. All package entries share the same package manager name and
+// install options.
+func MakePackageList(pm PackageManagerName, opts string, packages ...string) []Package {
+	var list []Package
+	for _, pkg := range packages {
+		list = append(list, Package{
+			Name:           pkg,
+			InstallOptions: opts,
+			PackageManager: pm,
+		})
+	}
+
+	return list
+}
+
+// InstallDependency executes the appropriate commands to install the specified
+// dependency targeting the provided series.
+func InstallDependency(dep Dependency, series string) error {
+	pkgManagers := make(map[PackageManagerName]manager.PackageManager)
+	pkgConfigurers := make(map[PackageManagerName]config.PackagingConfigurer)
+	pkgList, err := dep.PackageList(series)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, pkg := range pkgList {
+		pm := pkgManagers[pkg.PackageManager]
+		if pm == nil {
+			if pm, err = newPackageManager(pkg.PackageManager); err != nil {
+				return errors.Annotatef(err, "installing package %q via %q", pkg.Name, pkg.PackageManager)
+			}
+			pkgManagers[pkg.PackageManager] = pm
+		}
+
+		pkgConfer, exists := pkgConfigurers[pkg.PackageManager]
+		if !exists {
+			if pkgConfer, err = newPackageConfigurer(pkg.PackageManager, series); err != nil {
+				return errors.Annotatef(err, "installing package %q via %q", pkg.Name, pkg.PackageManager)
+			}
+			pkgConfigurers[pkg.PackageManager] = pkgConfer
+		}
+
+		if pkgConfer != nil {
+			if config.SeriesRequiresCloudArchiveTools(series) &&
+				pkgConfer.IsCloudArchivePackage(pkg.Name) {
+				extraOpts := pkgConfer.ApplyCloudArchiveTarget(pkg.Name)
+				pkg.appendInstallOptions(extraOpts[:len(extraOpts)-1]...)
+			}
+
+			if config.RequiresBackports(series, pkg.Name) {
```

Comment:
This should be moved into the Dependency, shouldn't it?

Reply:
Well... the current approach that we already use is a bit ugly. These config bits live in the juju/packaging repo and work on a per-package basis (see: https://github.com/juju/packaging/blob/aa6c1d0ba66b1630ed07f0345039fe13ba11f28a/config/ubuntu_constants.go#L90 for an example).

As the mongo/lxd/kvm install code included calls to the configurators it made sense to extract the logic and add it to a single place (the installer) instead of repeating it for each dependency.

Ideally, we should refactor (even better: rewrite) the configuration stuff but I didn't want to expand the scope of this PR further.

Reply:
I believe this is only used for Precise, which makes me think we can just drop it altogether.

Reply:
Ok then, I will update the PR and remove that.

Reply:
Removed the cloud tools check which are precise-only but kept the backports check as it is used by trusty. Once we stop supporting trusty I can revisit this and remove the configuration bits altogether.